<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Importing Libraries
<hr style="border-top:1px dashed">
 we will need the following libraries:

In [13]:
import pandas as pd
import numpy as np

In [22]:
import pandas as pd

df_test = pd.read_csv("data/2024-01/2024-01-essex-street.csv")
df_test.head()


,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context
0,NaN,2024-01,Essex Police,Essex Police,0.971553,51.963597,On or near The Street,E01029905,Babergh 010G,Anti-social behaviour,NaN,NaN
1,1ff5ad6a9707fbd93609bc66262d82a491be53ce9dd2bf...,2024-01,Essex Police,Essex Police,0.432091,51.640106,On or near Oakwood Drive,E01021237,Basildon 001A,Violence and sexual offences,Investigation complete; no suspect identified,NaN
2,1a5b1ab6ba14959b903fe6bd40db13d005a737b7f75b29...,2024-01,Essex Police,Essex Police,0.432091,51.640106,On or near Oakwood Drive,E01021237,Basildon 001A,Violence and sexual offences,Further action is not in the public interest,NaN
3,3ad3e4a02122cbb60a762d066a63c776d05d0b74594629...,2024-01,Essex Police,Essex Police,0.432091,51.640106,On or near Oakwood Drive,E01021237,Basildon 001A,Violence and sexual offences,Unable to prosecute suspect,NaN
4,4759e64e31495221e88d2a88b368ef81b29549a35852cc...,2024-01,Essex Police,Essex Police,0.435943,51.643479,On or near Penwood Close,E01021238,Basildon 001B,Violence and sexual offences,Investigation complete; no suspect identified,NaN


Next step: combining all the months and years for each police force

In [32]:
import pandas as pd

months = [
    "2024-01","2024-02","2024-03","2024-04","2024-05","2024-06","2024-07",
    "2024-08","2024-09","2024-10","2024-11","2024-12",
    "2025-01","2025-02","2025-03","2025-04","2025-05","2025-06","2025-07",
    "2025-08","2025-09","2025-10","2025-11","2025-12",
    "2026-01"
]

forces = ["city-of-london", "essex", "surrey", "west-midlands"]

df_list = []

for month in months:
    for force in forces:
        file_path = f"data/{month}/{month}-{force}-street.csv"
        try:
            temp = pd.read_csv(file_path)
            temp["force"] = force
            temp["year"] = month[:4]
            df_list.append(temp)
        except:
            pass



In [33]:
df = pd.concat(df_list, ignore_index=True)

In [35]:
df

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,force,year
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,NaN,city-of-london,2024
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,NaN,city-of-london,2024
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,NaN,city-of-london,2024
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,NaN,city-of-london,2024
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,NaN,city-of-london,2024
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1218618,5b27e6da01af3a9495e72731c87d35472073edb2e97721...,2026-01,West Midlands Police,West Midlands Police,-2.117079,52.571358,On or near Willcock Road,E01034315,Wolverhampton 035J,Violence and sexual offences,Under investigation,NaN,west-midlands,2026
1218619,0688659bb421c0a3c24c73954f90414e460581bdc0f509...,2026-01,West Midlands Police,West Midlands Police,-2.117890,52.570935,On or near Rowan Tree Drive,E01034315,Wolverhampton 035J,Violence and sexual offences,Under investigation,NaN,west-midlands,2026
1218620,02e3d5273fcf16f2e2bbd900549dfb39eed960cf712bbc...,2026-01,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Awaiting court outcome,NaN,west-midlands,2026
1218621,6c560e5b41678c83f1a2736d7e2019968867b72becad31...,2026-01,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN,west-midlands,2026


<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Initial Data Checks
<hr style="border-top:1px dashed">
 

Viewing first and last few rows to confirm structure:

In [43]:
df.head()

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,force,year
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,NaN,city-of-london,2024
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,NaN,city-of-london,2024
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,NaN,city-of-london,2024
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,NaN,city-of-london,2024
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,NaN,city-of-london,2024


In [45]:
df.tail()

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,force,year
1218618,5b27e6da01af3a9495e72731c87d35472073edb2e97721...,2026-01,West Midlands Police,West Midlands Police,-2.117079,52.571358,On or near Willcock Road,E01034315,Wolverhampton 035J,Violence and sexual offences,Under investigation,NaN,west-midlands,2026
1218619,0688659bb421c0a3c24c73954f90414e460581bdc0f509...,2026-01,West Midlands Police,West Midlands Police,-2.117890,52.570935,On or near Rowan Tree Drive,E01034315,Wolverhampton 035J,Violence and sexual offences,Under investigation,NaN,west-midlands,2026
1218620,02e3d5273fcf16f2e2bbd900549dfb39eed960cf712bbc...,2026-01,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Awaiting court outcome,NaN,west-midlands,2026
1218621,6c560e5b41678c83f1a2736d7e2019968867b72becad31...,2026-01,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Unable to prosecute suspect,NaN,west-midlands,2026
1218622,ed4f9a2b5ac1afc744506972e35ab9c69027bd0a055213...,2026-01,West Midlands Police,West Midlands Police,-2.118019,52.569505,On or near Thompson Avenue,E01034315,Wolverhampton 035J,Violence and sexual offences,Local resolution,NaN,west-midlands,2026


Seeing all column names to see which fields to clean


In [50]:
df.columns

Index(['Crime ID', 'Month', 'Reported by', 'Falls within', 'Longitude',
       'Latitude', 'Location', 'LSOA code', 'LSOA name', 'Crime type',
       'Last outcome category', 'Context', 'force', 'year'],
      dtype='object')

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Seeing how big the dataset is
<hr style="border-top:1px dashed">


In [53]:
df.shape

(1218623, 14)

In [88]:
df.info

<bound method DataFrame.info of                                                   Crime ID    Month  \
0        e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...  2024-01   
1        e60a5ac62a80e866453254474137c3206417422c62f0c0...  2024-01   
2        986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...  2024-01   
3        05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...  2024-01   
4        373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...  2024-01   
...                                                    ...      ...   
1218618  5b27e6da01af3a9495e72731c87d35472073edb2e97721...  2026-01   
1218619  0688659bb421c0a3c24c73954f90414e460581bdc0f509...  2026-01   
1218620  02e3d5273fcf16f2e2bbd900549dfb39eed960cf712bbc...  2026-01   
1218621  6c560e5b41678c83f1a2736d7e2019968867b72becad31...  2026-01   
1218622  ed4f9a2b5ac1afc744506972e35ab9c69027bd0a055213...  2026-01   

                   Reported by           Falls within  Longitude   Latitude  \
0        City of London Police  City

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Data types for each column
<hr style="border-top:1px dashed">

In [55]:
df.dtypes

Crime ID                  object
Month                     object
Reported by               object
Falls within              object
Longitude                float64
Latitude                 float64
Location                  object
LSOA code                 object
LSOA name                 object
Crime type                object
Last outcome category     object
Context                  float64
force                     object
year                      object
dtype: object

Month is currently a string so needs to be converted to datetime for a time series EDA

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Counting missing values per column
<hr style="border-top:1px dashed">



In [59]:
df.isna().sum()

Crime ID                  110568
Month                          0
Reported by                    0
Falls within                   0
Longitude                   9556
Latitude                    9556
Location                       0
LSOA code                   9556
LSOA name                   9556
Crime type                     0
Last outcome category     110568
Context                  1218623
force                          0
year                           0
dtype: int64

From here we can see there are 110568 crime IDs missing with the area langitude and latitude of 9556 with the LSOA code and name meaning within 110568 crimes, 9556 places are unidentified so will need to be handled. 110568 have no outcome category and 1218623 have missing contexts so 7 columns need to be cleaned.

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Unique Values for key columns
<hr style="border-top:1px dashed">

Unique Forces

In [79]:
df['force'].unique()

array(['city-of-london', 'essex', 'surrey', 'west-midlands'], dtype=object)

Unique Crime Types

In [86]:
df['Crime type'].unique()

array(['Theft from the person', 'Other theft', 'Burglary', 'Drugs',
       'Robbery', 'Violence and sexual offences', 'Anti-social behaviour',
       'Public order', 'Bicycle theft', 'Shoplifting',
       'Criminal damage and arson', 'Possession of weapons',
       'Vehicle crime', 'Other crime'], dtype=object)

Unique Last Outcome categories

In [96]:
df['Last outcome category'].unique()

array(['Investigation complete; no suspect identified',
       'Status update unavailable', 'Unable to prosecute suspect',
       'Formal action is not in the public interest',
       'Court result unavailable', nan,
       'Offender given a drugs possession warning',
       'Offender given a caution', 'Local resolution',
       'Action to be taken by another organisation',
       'Further investigation is not in the public interest',
       'Offender given penalty notice',
       'Further action is not in the public interest',
       'Awaiting court outcome',
       'Suspect charged as part of another case', 'Under investigation'],
      dtype=object)

Unique Contexts

In [99]:
df['Context'].unique()

array([nan])

The context column has the most null values and has no unique values meaning this column is always empty so this can be removed.

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Converting String Month to datetime
<hr style="border-top:1px dashed">

In [109]:
df["Month"] = pd.to_datetime(df["Month"])

Check month date type

In [119]:
df["Month"].dtype


dtype('<M8[ns]')

In [121]:
df["Month"].min(), df["Month"].max()


(Timestamp('2024-01-01 00:00:00'), Timestamp('2026-01-01 00:00:00'))

Confirm the datatype is datetime

In [126]:
df.dtypes

Crime ID                         object
Month                    datetime64[ns]
Reported by                      object
Falls within                     object
Longitude                       float64
Latitude                        float64
Location                         object
LSOA code                        object
LSOA name                        object
Crime type                       object
Last outcome category            object
Context                         float64
force                            object
year                             object
dtype: object

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Drop missing values : Context
<hr style="border-top:1px dashed">

Removing context column: this is always empty in the UK police data so it doesn't add any analytical value so this would simplify the dataset.

In [135]:
df.drop(columns=["Context"], inplace=True)


Confirm that context no longer exists in dataset

In [138]:
df.columns

Index(['Crime ID', 'Month', 'Reported by', 'Falls within', 'Longitude',
       'Latitude', 'Location', 'LSOA code', 'LSOA name', 'Crime type',
       'Last outcome category', 'force', 'year'],
      dtype='object')

In [140]:
df.isna().sum()

Crime ID                 110568
Month                         0
Reported by                   0
Falls within                  0
Longitude                  9556
Latitude                   9556
Location                      0
LSOA code                  9556
LSOA name                  9556
Crime type                    0
Last outcome category    110568
force                         0
year                          0
dtype: int64

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Drop missing rows: latitude/longitude
<hr style="border-top:1px dashed">

Rows with missing latitude or longitude cannot be mapped or spacially analysed and they also come with the LSOA code and LSOA name meaning removing latitude or longityde would remove the same rows where the LSOA code/name are missing

In [149]:
df = df[df["Latitude"].notna() & df["Longitude"].notna()]

In [151]:
df.isna().sum()

Crime ID                 110490
Month                         0
Reported by                   0
Falls within                  0
Longitude                     0
Latitude                      0
Location                      0
LSOA code                     0
LSOA name                     0
Crime type                    0
Last outcome category    110490
force                         0
year                          0
dtype: int64

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Filling Missing Data: Last outcome category
<hr style="border-top:1px dashed">

Sample data shows outcomes however the missing values can mean the case is still pending or not updated which is normal and outcomes can take months to be completed meaning the case could still be open so instead the missing values should be filled with outcome not recorded yet.

In [160]:
df.loc[:, "Last outcome category"] = df["Last outcome category"].fillna("Outcome not recorded yet")



In [164]:
df.isna().sum()

Crime ID                 110490
Month                         0
Reported by                   0
Falls within                  0
Longitude                     0
Latitude                      0
Location                      0
LSOA code                     0
LSOA name                     0
Crime type                    0
Last outcome category         0
force                         0
year                          0
dtype: int64

In [166]:
df['Last outcome category'].unique()

array(['Investigation complete; no suspect identified',
       'Status update unavailable', 'Unable to prosecute suspect',
       'Formal action is not in the public interest',
       'Court result unavailable', 'Outcome not recorded yet',
       'Offender given a drugs possession warning',
       'Offender given a caution', 'Local resolution',
       'Action to be taken by another organisation',
       'Further investigation is not in the public interest',
       'Offender given penalty notice',
       'Further action is not in the public interest',
       'Awaiting court outcome',
       'Suspect charged as part of another case', 'Under investigation'],
      dtype=object)

<a name="anchorInstall" style="position:absolute;"></a>
<hr style="border:2px solid">

# Filling Missing Data: Crime ID
<hr style="border-top:1px dashed">

Crime ID has a lot of missing values but many crimes can be anonymised or withheld by polce forces meaning missing values does not mean the data is unreliable but it means simply the ID wasn't provided therefore it should be filled.

In [176]:
df.loc[:, "Crime ID"] = df["Crime ID"].fillna("No Crime ID")


In [180]:
df["Crime ID"].isna().sum()

0

In [182]:
df.isna().sum()

Crime ID                 0
Month                    0
Reported by              0
Falls within             0
Longitude                0
Latitude                 0
Location                 0
LSOA code                0
LSOA name                0
Crime type               0
Last outcome category    0
force                    0
year                     0
dtype: int64

Next save the clean dataset to then work on the Analysis notebook

In [ ]:
df.to_csv("cleaned_crime_data.csv", index=False)